### PR TITLE
Fixes #17841 - Expose compute resource SSH keys via API

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

File: app/views/api/v2/key_pairs/main.json.rabl [hunk 1]
```diff
@@ -0,0 +1,4 @@
+object @key_pair
+
+attributes :name, :active, :used_elsewhere
```
Comment (ohadlevy) [inline app/views/api/v2/key_pairs/main.json.rabl]: maybe in_used vs used_elsewhere?
is there a way via the API / UI to figure out where its being used?

File: app/controllers/api/v2/key_pairs_controller.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        process_response @compute_resource.recreate
+      end
+
+      api :DELETE, '/compute_resources/:compute_resource_id/key_pairs/:id', N_("Show SSH key pair of a compute resource")
+      param :compute_resource_id, String, :required => true, :desc => N_("ID or name of compute resource")
+      param :id, String, :required => true, :desc => N_("ID of SSH key pair")
+
+      def destroy
+        key_to_delete = params[:id]
+        return not_found unless key_to_delete
```
Comment (ohadlevy) [inline app/controllers/api/v2/key_pairs_controller.rb:37]: shouldnt this be recreate vs delete? I mean we should allow to delete unused keys, but the current active key should not be deleted only recreated?

File: app/views/api/v2/key_pairs/main.json.rabl (hunk 1 above)
Comment (shlomizadok) [reply app/views/api/v2/key_pairs/main.json.rabl]: I'll reword. Currently no way to know where it is used

File: app/controllers/api/v2/key_pairs_controller.rb (hunk 2 above)
Comment (shlomizadok) [reply app/controllers/api/v2/key_pairs_controller.rb:37]: Yes. This should be checked at the model. will refactor.

**Comment by shlomizadok** [comment]
[test]

**Comment by tbrisker** [comment]
[test foreman]
[test katello]
[test upgrade]

File: app/controllers/api/v2/key_pairs_controller.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+module Api
+  module V2
+    class KeyPairsController <  V2::BaseController
+      include Api::Version2
+
+      before_action :find_compute_resource, :except => [:show]
+      before_action :find_resource, :only => [:show]
+
+      api :GET, '/compute_resources/:compute_resource_id/key_pairs', N_("List all SSH key for compute resource")
```
Comment (dLobatog) [inline app/controllers/api/v2/key_pairs_controller.rb:9]: typo, 'SSH key pairs` (can be fixed on merge)

File: app/controllers/api/v2/key_pairs_controller.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+      private
+
+      def find_compute_resource
+        @compute_resource = ComputeResource.find(params[:compute_resource_id])
+        return not_found unless @compute_resource.capabilities.include?(:key_pair)
+        @compute_resource
+      end
+
+      def find_resource
```
Comment (dLobatog) [inline app/controllers/api/v2/key_pairs_controller.rb:49]: No need to override this method?

File: app/controllers/api/v2/key_pairs_controller.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        @key_pair = KeyPair.find(params[:id])
+      end
+
+      def get_resource
```
Comment (dLobatog) [inline app/controllers/api/v2/key_pairs_controller.rb:53]: Dead code?

File: app/controllers/api/v2/key_pairs_controller.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+      def destroy
+        key_to_delete = params[:id]
+        return not_found unless key_to_delete
+        process_response @compute_resource.delete_key_pair(key_to_delete)
+      end
+
+      private
+
+      def find_compute_resource
```
Comment (dLobatog) [inline app/controllers/api/v2/key_pairs_controller.rb:43]: This looks like it should rely on `find_required_nested_object`, then check whether @compute_resource.capabilities include keypair?

File: app/controllers/api/v2/key_pairs_controller.rb (hunk 2 above)
Comment (dLobatog) [reply app/controllers/api/v2/key_pairs_controller.rb:37]: I see the refactor allows to destroy "unused keys" defining "unused" as "not used by Foreman". I'd say that looks a bit dangerous.

**Review by dLobatog — CHANGES_REQUESTED** [review]
@shlomizadok Overall it works nicely , found a few things in the controller that I think it could reuse from the API ApplicationController - mind to check it out. Thanks!

File: app/controllers/api/v2/key_pairs_controller.rb (hunk 6 above)
Comment (shlomizadok) [reply app/controllers/api/v2/key_pairs_controller.rb:43]: It fails when relying on `find_required_nested_object`

File: app/controllers/api/v2/key_pairs_controller.rb (hunk 5 above)
Comment (shlomizadok) [reply app/controllers/api/v2/key_pairs_controller.rb:53]: nope

**Comment by shlomizadok** [comment]
@dLobatog - how shall I proceed here? :)

**Comment by ohadlevy** [comment]
@dLobatog recommendations on how to continue?

**Pull request closed without merging**